sunkit_image/granule.py

```diff
@@ -0,0 +1,289 @@
+import scipy
+import skimage
+from sklearn.cluster import KMeans
+import matplotlib
+
+import sunpy
+import sunpy.map
+
+__all__ = [
+    "segment"
+]
+
+
+def segment(smap, *, skimage_method="li", mark_dim_centers=False, bp_min_flux=None):
+    """
+    Segment an optical image of the solar photosphere into tri-value maps with:
+
+     * 0 as intergranule
+     * 1 as granule
+     * 2 as brightpoint 
+
+    If mark_dim_centers is set to True, an additional label, 3, will be assigned to
+    dim grnanule centers. 
+
+    Parameters
+    ----------
+    smap : `~sunpy.map.GenericMap`
+        `~sunpy.map.GenericMap` containing data to segment. Must have square pixels.
+    skimage_method : {"li", "otsu", "isodata", "mean", "minimum", "yen", "triangle"}, optional
+        scikit-image thresholding method, defaults to "li".
+        Depending on input data, one or more of these methods may be
+        significantly better or worse than the others. Typically, 'li', 'otsu',
+        'mean', and 'isodata' are good choices, 'yen' and 'triangle' over-
+        identify intergranule material, and 'minimum' over identifies granules.
+    mark_dim_centers : `bool`, optional
+        Whether to mark dim granule centers as a separate category for future exploration.
+    bp_min_flux : `float`, optional
+        Minimum flux per pixel for a region to be considered a brightpoint.
+        Default is `None` which will use data mean + 0.5 * sigma.
+
+    Returns
+    -------
+    segmented_map : `~sunpy.map.GenericMap`
+        `~sunpy.map.GenericMap` containing a segmented image (with the original header).
+    """
+    if not isinstance(smap, sunpy.map.mapbase.GenericMap):
+        raise TypeError("Input must be an instance of a sunpy.map.GenericMap")
+
+    if map.scale[0].value == map.scale[1].value:
+        resolution = map.scale[0].value
+    else: 
+        raise Exception('Currently only maps with square pixels are supported.')
+
+    median_filtered = scipy.ndimage.median_filter(smap.data, size=3)
+    # Apply initial skimage threshold.
+    threshold = _get_threshold(median_filtered, skimage_method)
+    segmented_image = np.uint8(median_filtered > threshold)
+    # Fix the extra intergranule material bits in the middle of granules.
+    seg_im_fixed = _trim_intergranules(segmented_image, mark=mark_dim_centers)
+    # Mark brightpoint and get final granule and brightpoint count.
+    seg_im_markbp, brightpoint_count, granule_count = _mark_brightpoint(seg_im_fixed, smap.data, resolution, bp_min_flux)
+    logging.info(f"Segmentation has identified {granule_count} granules and {brightpoint_count} brightpoint")
+    segmented_map = sunpy.map.Map(seg_im_markbp, smap.meta)
+    # Add colormap to Map, such that 0 (intergranules) = black, 1 (granule) = white, 2 (brightpoints) = yellow, 3 (dim_centers) = blue
+    cmap = matplotlib.colors.ListedColormap(["black", "white", "#ffc406", "blue"])
+    segmented_map.plot_settings['cmap'] = cmap

```

Comment:
Adding an appropriate norm will make sure the colors fall in the right place,

```suggestion
    cmap = matplotlib.colors.ListedColormap(["black", "white", "#ffc406", "blue"])
    norm = matplotlib.colors.BoundaryNorm(boundaries=[-.5, .5, 1.5, 2.5, 3.5], ncolors=cmap.N)
    segmented_map.plot_settings['cmap'] = cmap
    segmented_map.plot_settings['norm'] = norm
```

As an example, for the following array,
```
[[2, 1, 0],
 [3, 2, 3],
 [1, 0, 2]]
```
the above norm and colormap give the following image,
```python
im = plt.imshow(foo,cmap=cmap,norm=norm)
cb = plt.colorbar(im, ticks=[0,1,2,3])
cb.ax.set_yticklabels(['intergranule', 'granule', 'brightpoint', 'dim granule'])
```
![image](https://user-images.githubusercontent.com/8676141/215126497-ac4e60a5-2715-4b8b-ab4b-759b0ebdb111.png)

Reply:
perfect!